File: windows/Hinge.Core/PairingManager.cs
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json.Serialization;

namespace Hinge.Core;

public class PairRequestMessage
{
    [JsonPropertyName("initiatorDeviceId")]
    public string InitiatorDeviceId { get; set; } = string.Empty;

    [JsonPropertyName("initiatorName")]
    public string InitiatorName { get; set; } = string.Empty;

    [JsonPropertyName("receiverDeviceId")]
    public string ReceiverDeviceId { get; set; } = string.Empty;

    [JsonPropertyName("salt")]
    public string Salt { get; set; } = string.Empty;

    [JsonPropertyName("initiatorNonce")]
    public string InitiatorNonce { get; set; } = string.Empty;

    [JsonPropertyName("initiatorPublicKey")]
    public string InitiatorPublicKey { get; set; } = string.Empty;

    [JsonPropertyName("timestamp")]
    public long Timestamp { get; set; } = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
}

public class PairConfirmMessage
{
    [JsonPropertyName("initiatorDeviceId")]
    public string InitiatorDeviceId { get; set; } = string.Empty;

    [JsonPropertyName("receiverDeviceId")]
    public string ReceiverDeviceId { get; set; } = string.Empty;

    [JsonPropertyName("receiverName")]
    public string ReceiverName { get; set; } = string.Empty;

    [JsonPropertyName("receiverNonce")]
    public string ReceiverNonce { get; set; } = string.Empty;

    [JsonPropertyName("receiverPublicKey")]
    public string ReceiverPublicKey { get; set; } = string.Empty;

    [JsonPropertyName("sasCode")]
    public string SasCode { get; set; } = string.Empty;

    [JsonPropertyName("accepted")]
    public bool Accepted { get; set; }

    [JsonPropertyName("timestamp")]
    public long Timestamp { get; set; } = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
}

public class PairingManager
{
    private readonly DeviceIdentity _localIdentity;
    private readonly TrustStore _trustStore;

    public PairingManager(DeviceIdentity localIdentity, TrustStore trustStore)
    {
        _localIdentity = localIdentity;
        _trustStore = trustStore;
    }

    /// <summary>
    /// Backward-compatible legacy deterministic PIN derivation (prototype).
    /// </summary>
    public static int DerivePin(string initiatorId, string receiverId, string salt)
    {
        string input = $"{initiatorId}:{receiverId}:{salt}";
        byte[] hash = SHA256.HashData(Encoding.UTF8.GetBytes(input));
        uint val = BinaryPrimitives.ReadUInt32BigEndian(hash.AsSpan(0, 4));
        return (int)((val % 900000) + 100000);
    }

    /// <summary>
    /// Deterministic 6-digit Short Authentication String (SAS) derivation:
    /// HMAC_SHA256(Secret, "Hinge-SAS-v1" || NonceA || NonceB || PubKeyA || PubKeyB)
    /// </summary>
    public static string DeriveSasCode(byte[] sharedSecret, byte[] nonceA, byte[] nonceB, string pubKeyA = "", string pubKeyB = "")
    {
        byte[] prefix = Encoding.UTF8.GetBytes("Hinge-SAS-v1");
        byte[] pubKeyABytes = Encoding.UTF8.GetBytes(pubKeyA);
        byte[] pubKeyBBytes = Encoding.UTF8.GetBytes(pubKeyB);

        int totalLen = prefix.Length + nonceA.Length + nonceB.Length + pubKeyABytes.Length + pubKeyBBytes.Length;
        byte[] context = new byte[totalLen];
        int offset = 0;
        Buffer.BlockCopy(prefix, 0, context, offset, prefix.Length);
        offset += prefix.Length;
        Buffer.BlockCopy(nonceA, 0, context, offset, nonceA.Length);
        offset += nonceA.Length;
        Buffer.BlockCopy(nonceB, 0, context, offset, nonceB.Length);
        offset += nonceB.Length;
        Buffer.BlockCopy(pubKeyABytes, 0, context, offset, pubKeyABytes.Length);
        offset += pubKeyABytes.Length;
        Buffer.BlockCopy(pubKeyBBytes, 0, context, offset, pubKeyBBytes.Length);

        using var hmac = new HMACSHA256(sharedSecret);
        byte[] digest = hmac.ComputeHash(context);

        uint truncatedInt = BinaryPrimitives.ReadUInt32BigEndian(digest.AsSpan(0, 4));
        uint code = (truncatedInt % 900000) + 100000;
        return code.ToString("D6");
    }

    public static string DeriveSasCode(string sharedSecretHexOrStr, string nonceAHex, string nonceBHex, string pubKeyA = "", string pubKeyB = "")
    {
        byte[] secret = Encoding.UTF8.GetBytes(sharedSecretHexOrStr);
        byte[] nA = Convert.FromHexString(nonceAHex);
        byte[] nB = Convert.FromHexString(nonceBHex);
        return DeriveSasCode(secret, nA, nB, pubKeyA, pubKeyB);
    }

    public PairRequestMessage CreatePairRequest(string receiverDeviceId, string? publicKey = null)
    {
        byte[] nonceBytes = new byte[16];
        RandomNumberGenerator.Fill(nonceBytes);
        string nonceHex = Convert.ToHexString(nonceBytes).ToLowerInvariant();

        return new PairRequestMessage
        {
            InitiatorDeviceId = _localIdentity.DeviceId,
            InitiatorName = _localIdentity.Name,
            ReceiverDeviceId = receiverDeviceId,
            Salt = nonceHex, // For backwards compatibility
            InitiatorNonce = nonceHex,
            InitiatorPublicKey = publicKey ?? _localIdentity.PublicKey,
            Timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
        };
    }

    public PairConfirmMessage CreatePairConfirm(PairRequestMessage request, bool accepted, byte[]? sharedSecret = null)
    {
        byte[] receiverNonceBytes = new byte[16];
        RandomNumberGenerator.Fill(receiverNonceBytes);
        string receiverNonceHex = Convert.ToHexString(receiverNonceBytes).ToLowerInvariant();

        string sasCode = string.Empty;
        if (accepted)
        {
            byte[] secret = sharedSecret ?? SHA256.HashData(Encoding.UTF8.GetBytes($"{request.InitiatorDeviceId}:{_localIdentity.DeviceId}"));
            byte[] nA = Convert.FromHexString(string.IsNullOrEmpty(request.InitiatorNonce) ? request.Salt.PadRight(32, '0')[..32] : request.InitiatorNonce);
            sasCode = DeriveSasCode(secret, nA, receiverNonceBytes, request.InitiatorPublicKey, _localIdentity.PublicKey);
        }

        return new PairConfirmMessage
        {
            InitiatorDeviceId = request.InitiatorDeviceId,
            ReceiverDeviceId = _localIdentity.DeviceId,
            ReceiverName = _localIdentity.Name,
            ReceiverNonce = receiverNonceHex,
            ReceiverPublicKey = _localIdentity.PublicKey,
            SasCode = sasCode,
            Accepted = accepted,
            Timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds()
        };
    }

    public void SaveTrustedPeer(string peerDeviceId, string peerName, string publicKey = "")
    {
        _trustStore.AddOrUpdate(new TrustedDevice
        {
            DeviceId = peerDeviceId,
            Name = peerName,
            PublicKey = publicKey,
            PairedAt = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
            LastSeen = DateTimeOffset.UtcNow.ToUnixTimeSeconds(),
            TrustState = TrustState.Trusted
        });
    }

    public bool IsPeerTrusted(string peerDeviceId)
    {
        return _trustStore.IsTrusted(peerDeviceId);
    }
}

File: android/lib/core/pairing_manager.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';

import 'device_identity_manager.dart';
import 'device_model.dart';
import 'trust_store.dart';

class PairRequestMessage {
  final String initiatorDeviceId;
  final String initiatorName;
  final String receiverDeviceId;
  final String salt;
  final String initiatorNonce;
  final String initiatorPublicKey;
  final int timestamp;

  const PairRequestMessage({
    required this.initiatorDeviceId,
    required this.initiatorName,
    required this.receiverDeviceId,
    required this.salt,
    this.initiatorNonce = '',
    this.initiatorPublicKey = '',
    required this.timestamp,
  });

  Map<String, dynamic> toJson() => {
    'initiatorDeviceId': initiatorDeviceId,
    'initiatorName': initiatorName,
    'receiverDeviceId': receiverDeviceId,
    'salt': salt,
    'initiatorNonce': initiatorNonce,
    'initiatorPublicKey': initiatorPublicKey,
    'timestamp': timestamp,
  };

  factory PairRequestMessage.fromJson(Map<String, dynamic> json) {
    return PairRequestMessage(
      initiatorDeviceId: json['initiatorDeviceId'] as String,
      initiatorName: json['initiatorName'] as String,
      receiverDeviceId: json['receiverDeviceId'] as String,
      salt: json['salt'] as String,
      initiatorNonce: json['initiatorNonce'] as String? ?? '',
      initiatorPublicKey: json['initiatorPublicKey'] as String? ?? '',
      timestamp: json['timestamp'] as int? ?? 0,
    );
  }
}

class PairConfirmMessage {
  final String initiatorDeviceId;
  final String receiverDeviceId;
  final String receiverName;
  final String receiverNonce;
  final String receiverPublicKey;
  final String sasCode;
  final bool accepted;
  final int timestamp;

  const PairConfirmMessage({
    required this.initiatorDeviceId,
    required this.receiverDeviceId,
    this.receiverName = '',
    this.receiverNonce = '',
    this.receiverPublicKey = '',
    this.sasCode = '',
    required this.accepted,
    required this.timestamp,
  });

  Map<String, dynamic> toJson() => {
    'initiatorDeviceId': initiatorDeviceId,
    'receiverDeviceId': receiverDeviceId,
    'receiverName': receiverName,
    'receiverNonce': receiverNonce,
    'receiverPublicKey': receiverPublicKey,
    'sasCode': sasCode,
    'accepted': accepted,
    'timestamp': timestamp,
  };

  factory PairConfirmMessage.fromJson(Map<String, dynamic> json) {
    return PairConfirmMessage(
      initiatorDeviceId: json['initiatorDeviceId'] as String,
      receiverDeviceId: json['receiverDeviceId'] as String,
      receiverName: json['receiverName'] as String? ?? '',
      receiverNonce: json['receiverNonce'] as String? ?? '',
      receiverPublicKey: json['receiverPublicKey'] as String? ?? '',
      sasCode: json['sasCode'] as String? ?? '',
      accepted: json['accepted'] as bool? ?? false,
      timestamp: json['timestamp'] as int? ?? 0,
    );
  }
}

class PairingManager {
  final DeviceIdentity _localIdentity;
  final TrustStore _trustStore;

  PairingManager({required this._localIdentity, required this._trustStore});

  static String normalizePairingCode(String? value) {
    final code = value?.trim() ?? '';
    return RegExp(r'^\d{6}$').hasMatch(code) ? code : '';
  }

  static bool isValidPairingCode(String? value) =>
      normalizePairingCode(value).length == 6;

  static String createChallenge() {
    final random = Random.secure();
    return List<int>.generate(16, (_) => random.nextInt(256))
        .map((byte) => byte.toRadixString(16).padLeft(2, '0'))
        .join();
  }

  static String createProof(String pairingCode, String challenge) {
    final normalized = normalizePairingCode(pairingCode);
    if (normalized.isEmpty || challenge.trim().isEmpty) return '';
    final key = sha256.convert(utf8.encode(normalized)).bytes;
    final context = utf8.encode(
      'Hinge-Pairing-v1|${challenge.trim().toLowerCase()}',
    );
    return Hmac(sha256, key).convert(context).toString().toLowerCase();
  }

  static bool verifyProof(
    String pairingCode,
    String challenge,
    String proof,
  ) {
    final expected = createProof(pairingCode, challenge);
    if (expected.isEmpty || proof.trim().isEmpty) return false;
    try {
      final expectedBytes = _hexBytes(expected);
      final actualBytes = _hexBytes(proof.trim());
      if (expectedBytes.length != actualBytes.length) return false;
      var difference = 0;
      for (var index = 0; index < expectedBytes.length; index++) {
        difference |= expectedBytes[index] ^ actualBytes[index];
      }
      return difference == 0;
    } catch (_) {
      return false;
    }
  }

  static List<int> _hexBytes(String value) {
    if (value.length.isOdd) throw const FormatException('invalid hex');
    return [
      for (var index = 0; index < value.length; index += 2)
        int.parse(value.substring(index, index + 2), radix: 16),
    ];
  }

  /// Backward-compatible legacy deterministic PIN derivation (prototype).
  static int derivePin(String initiatorId, String receiverId, String salt) {
    final input = '$initiatorId:$receiverId:$salt';
    final bytes = utf8.encode(input);
    final digest = sha256.convert(bytes);
    final byteData = ByteData.sublistView(
      Uint8List.fromList(digest.bytes.sublist(0, 4)),
    );
    final val = byteData.getUint32(0, Endian.big);
    return (val % 900000) + 100000;
  }

  /// Deterministic 6-digit Short Authentication String (SAS) derivation:
  /// HMAC_SHA256(Secret, "Hinge-SAS-v1" || NonceA || NonceB || PubKeyA || PubKeyB)
  static String deriveSasCode({
    required List<int> sharedSecret,
    required List<int> nonceA,
    required List<int> nonceB,
    String pubKeyA = '',
    String pubKeyB = '',
  }) {
    final prefix = utf8.encode('Hinge-SAS-v1');
    final pA = utf8.encode(pubKeyA);
    final pB = utf8.encode(pubKeyB);

    final context = <int>[...prefix, ...nonceA, ...nonceB, ...pA, ...pB];

    final hmac = Hmac(sha256, sharedSecret);
    final digest = hmac.convert(context);
    final byteData = ByteData.sublistView(
      Uint8List.fromList(digest.bytes.sublist(0, 4)),
    );
    final truncatedInt = byteData.getUint32(0, Endian.big);
    final code = (truncatedInt % 900000) + 100000;
    return code.toString().padLeft(6, '0');
  }

  static String deriveSasCodeFromHex({
    required String sharedSecretStr,
    required String nonceAHex,
    required String nonceBHex,
    String pubKeyA = '',
    String pubKeyB = '',
  }) {
    final secret = utf8.encode(sharedSecretStr);
    final nA = _parseHex(nonceAHex);
    final nB = _parseHex(nonceBHex);
    return deriveSasCode(
      sharedSecret: secret,
      nonceA: nA,
      nonceB: nB,
      pubKeyA: pubKeyA,
      pubKeyB: pubKeyB,
    );
  }

  static List<int> _parseHex(String hex) {
    final result = <int>[];
    for (var i = 0; i < hex.length - 1; i += 2) {
      result.add(int.parse(hex.substring(i, i + 2), radix: 16));
    }
    return result;
  }

  PairRequestMessage createPairRequest(
    String receiverDeviceId, {
    String? publicKey,
  }) {
    final rand = Random.secure();
    final nonceBytes = List<int>.generate(16, (_) => rand.nextInt(256));
    final nonceHex = nonceBytes
        .map((b) => b.toRadixString(16).padLeft(2, '0'))
        .join();

    return PairRequestMessage(
      initiatorDeviceId: _localIdentity.deviceId,
      initiatorName: _localIdentity.name,
      receiverDeviceId: receiverDeviceId,
      salt: nonceHex,
      initiatorNonce: nonceHex,
      initiatorPublicKey: publicKey ?? _localIdentity.publicKey,
      timestamp: DateTime.now().millisecondsSinceEpoch ~/ 1000,
    );
  }

  PairConfirmMessage createPairConfirm(
    PairRequestMessage request, {
    required bool accepted,
    List<int>? sharedSecret,
  }) {
    final rand = Random.secure();
    final receiverNonceBytes = List<int>.generate(16, (_) => rand.nextInt(256));
    final receiverNonceHex = receiverNonceBytes
        .map((b) => b.toRadixString(16).padLeft(2, '0'))
        .join();

    String sasCode = '';
    if (accepted) {
      final secret =
          sharedSecret ??
          sha256
              .convert(
                utf8.encode(
                  '${request.initiatorDeviceId}:${_localIdentity.deviceId}',
                ),
              )
              .bytes;
      final nAHex = request.initiatorNonce.isNotEmpty
          ? request.initiatorNonce
          : request.salt.padRight(32, '0').substring(0, 32);
      final nA = _parseHex(nAHex);
      sasCode = deriveSasCode(
        sharedSecret: secret,
        nonceA: nA,
        nonceB: receiverNonceBytes,
        pubKeyA: request.initiatorPublicKey,
        pubKeyB: _localIdentity.publicKey,
      );
    }

    return PairConfirmMessage(
      initiatorDeviceId: request.initiatorDeviceId,
      receiverDeviceId: _localIdentity.deviceId,
      receiverName: _localIdentity.name,
      receiverNonce: receiverNonceHex,
      receiverPublicKey: _localIdentity.publicKey,
      sasCode: sasCode,
      accepted: accepted,
      timestamp: DateTime.now().millisecondsSinceEpoch ~/ 1000,
    );
  }

  void saveTrustedPeer(
    String peerDeviceId,
    String peerName, {
    String publicKey = '',
  }) {
    _trustStore.addOrUpdate(
      TrustedDevice(
        deviceId: peerDeviceId,
        name: peerName,
        publicKey: publicKey,
        pairedAt: DateTime.now().millisecondsSinceEpoch ~/ 1000,
        lastSeen: DateTime.now().millisecondsSinceEpoch ~/ 1000,
        trustState: DeviceTrustState.trusted,
      ),
    );
  }

  bool isPeerTrusted(String peerDeviceId) {
    return _trustStore.isTrusted(peerDeviceId);
  }
}
